File: WordPress/Classes/Utility/ContextManager.m
#import "ContextManager.h"
#import "LegacyContextFactory.h"
#import "WordPress-Swift.h"
@import WordPressShared.WPAnalytics;
@import Foundation;

#define SentryStartupEventAddError(event, error) [event addError:error file:__FILE__ function:__FUNCTION__ line:__LINE__]

NSString * const ContextManagerModelNameCurrent = @"$CURRENT";

// MARK: - Static Variables
//
static ContextManager *_instance;


// MARK: - Private Properties
//
@interface ContextManager ()

@property (nonatomic, strong) NSPersistentStoreDescription *storeDescription;
@property (nonatomic, strong) NSPersistentContainer *persistentContainer;
@property (nonatomic, strong) NSManagedObjectModel *managedObjectModel;
@property (nonatomic, assign) BOOL migrationFailed;
@property (nonatomic, strong) NSString *modelName;
@property (nonatomic, strong) NSURL *storeURL;
@property (nonatomic, strong) id<ManagedObjectContextFactory> contextFactory;
@property (nonatomic, strong) NSOperationQueue *remoteChangeQueue;
@property (nonatomic, strong) NSPersistentHistoryToken *historyToken;
@property (nonatomic, strong) id updateUIObserver;

@end


// MARK: - ContextManager
//
@implementation ContextManager

- (instancetype)init
{
    NSURL *storeURL;
    if ([self isSharedLoginEnabled]) {
        [self copyDatabaseToSharedDirectoryIfNeeded];
        storeURL = [self sharedDatabasePath];
    } else {
        storeURL = [self localDatabasePath];
    }

    return [self initWithModelName:ContextManagerModelNameCurrent storeURL:storeURL contextFactory:nil];
}

- (instancetype)initWithModelName:(NSString *)modelName storeURL:(NSURL *)storeURL contextFactory:(Class)factory
{
    self = [super init];
    if (self) {
        if (factory == nil) {
            factory = [Feature enabled:FeatureFlagNewCoreDataContext] ? [ContainerContextFactory class] : [LegacyContextFactory class];
        }

        NSParameterAssert([modelName isEqualToString:ContextManagerModelNameCurrent] || [modelName hasPrefix:@"WordPress "]);
        NSParameterAssert([storeURL isFileURL]);
        NSParameterAssert([factory conformsToProtocol:@protocol(ManagedObjectContextFactory)]);

        self.modelName = modelName;
        self.storeURL = storeURL;

        [NSValueTransformer registerCustomTransformers];
        self.contextFactory = (id<ManagedObjectContextFactory>)[[factory alloc] initWithPersistentContainer:self.persistentContainer];
        [[[NullBlogPropertySanitizer alloc] initWithContext:self.contextFactory.mainContext] sanitize];
        
        if ([self isSharedLoginEnabled]) {
            [self observeStoreChanges];
        }
    }

    return self;
}

+ (instancetype)internalSharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[ContextManager alloc] init];
    });

    return _instance;
}

+ (NSURL *)inMemoryStoreURL
{
    return [NSURL fileURLWithPath:@"/dev/null"];
}

#pragma mark - Contexts

- (NSManagedObjectContext *)mainContext
{
    return self.contextFactory.mainContext;
}

- (NSManagedObjectContext *const)newDerivedContext
{
    return [self.contextFactory newDerivedContext];
}

#pragma mark - Context Saving and Merging

- (void)saveContextAndWait:(NSManagedObjectContext *)context
{
    [self.contextFactory saveContext:context andWait:YES withCompletionBlock:nil];
}

- (void)saveContext:(NSManagedObjectContext *)context
{
    [self.contextFactory saveContext:context andWait:NO withCompletionBlock:nil];
}

- (void)saveContext:(NSManagedObjectContext *)context withCompletionBlock:(void (^)(void))completionBlock
{
    [self.contextFactory saveContext:context andWait:NO withCompletionBlock:completionBlock];
}

- (void)performAndSaveUsingBlock:(void (^)(NSManagedObjectContext *context))aBlock
{
    NSManagedObjectContext *context = [self newDerivedContext];
    [context performBlockAndWait:^{
        aBlock(context);

        [self saveContextAndWait:context];
    }];
}

- (void)performAndSaveUsingBlock:(void (^)(NSManagedObjectContext *context))aBlock completion:(void (^)(void))completion
{
    NSManagedObjectContext *context = [self newDerivedContext];
    [context performBlock:^{
        aBlock(context);

        [self.contextFactory saveContext:context andWait:NO withCompletionBlock:completion];
    }];
}

#pragma mark - Store Changes

- (void)observeStoreChanges
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleStoreChanges:)
                                                 name:NSPersistentStoreRemoteChangeNotification
                                               object:self.persistentContainer.persistentStoreCoordinator];
}

- (void)handleStoreChanges:(NSNotification *)notification
{
    NSLog(@"[CM]: Persistent store remote change notification");
    __weak __typeof(self) weakSelf = self;
    [self.remoteChangeQueue addOperationWithBlock:^{
        [weakSelf proccessRemoteChanges];
    }];
}

- (void)proccessRemoteChanges
{
    NSLog(@"[CM]: Processing remote changes");
    NSManagedObjectContext *context = [self newDerivedContext];
    [context performBlockAndWait:^{
        NSPersistentHistoryChangeRequest *historyChangeRequest = [NSPersistentHistoryChangeRequest fetchHistoryAfterToken:self.historyToken];
        NSFetchRequest *fetchRequest = [NSPersistentHistoryTransaction fetchRequest];
        fetchRequest.predicate = [NSPredicate predicateWithFormat:@"bundleID != %@", [[NSBundle mainBundle] bundleIdentifier]];
        historyChangeRequest.fetchRequest = fetchRequest;
        NSPersistentHistoryResult *result = [context executeRequest:historyChangeRequest error:nil];
        NSArray<NSPersistentHistoryTransaction *> *transactions = result.result;
        
        if (transactions.count > 0) {
            for (NSPersistentHistoryTransaction *transaction in transactions) {
                NSDictionary *userInfo = transaction.objectIDNotification.userInfo;
                
                if (userInfo) {
                    NSLog(@"[CM]: Merging into main context %@", userInfo);
                    [NSManagedObjectContext mergeChangesFromRemoteContextSave:userInfo
                                                                 intoContexts:@[self.mainContext]];
                }
            }
            [self saveContext:self.mainContext];
            self.historyToken = transactions.lastObject.token;
            dispatch_async(dispatch_get_main_queue(), ^{
                [self handleUIUpdate];
            });
        } else {
            NSLog(@"[CM]: No transactions");
        }
    
    }];
}

- (void)handleUIUpdate
{
    UIApplicationState state = UIApplication.sharedApplication.applicationState;
    
    void (^showUI)(void) = ^{
        NSLog(@"[CM]: Updating UI if necessary");

        WordPressAppDelegate *delegate = (WordPressAppDelegate *) UIApplication.sharedApplication.delegate;
        if (AccountHelper.isLoggedIn) {
            [WPTabBarController.sharedInstance reloadSplitViewControllers];
            [delegate.windowManager showAppUIFor:nil completion:nil];
        } else {
            [delegate.windowManager showFullscreenSignIn];
        }
    };
    if (state == UIApplicationStateActive) {
        showUI();
    } else if (!self.updateUIObserver) {
        __weak __typeof(self) weakSelf = self;
        self.updateUIObserver = [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationDidBecomeActiveNotification
                                                                                object:nil
                                                                                 queue:[NSOperationQueue mainQueue]
                                                                            usingBlock:^(NSNotification * _Nonnull note) {
            showUI();
            [[NSNotificationCenter defaultCenter] removeObserver:weakSelf.updateUIObserver];
            weakSelf.updateUIObserver = nil;
        }];
    }
}

#pragma mark - Setup

- (NSPersistentContainer *)persistentContainer
{
    if (_persistentContainer) {
        return _persistentContainer;
    }

    SentryStartupEvent *startupEvent = [SentryStartupEvent new];

    [self migrateDataModelsIfNecessary:startupEvent];

    NSURL *storeURL = self.storeURL;

    // Initialize the container
    NSPersistentContainer *persistentContainer = [[NSPersistentContainer alloc] initWithName:@"WordPress" managedObjectModel:self.managedObjectModel];
    _persistentContainer = persistentContainer;
    persistentContainer.persistentStoreDescriptions = @[self.storeDescription];
    [persistentContainer loadPersistentStoresWithCompletionHandler:^(NSPersistentStoreDescription *description, NSError *error) {
        if (error != nil) {
            DDLogError(@"Error opening the database. %@\nDeleting the file and trying again", error);

            SentryStartupEventAddError(startupEvent, error);
            error = nil;

            // make a backup of the old database
            [CoreDataIterativeMigrator backupDatabaseAt:storeURL error:&error];
            if (error != nil) {
                SentryStartupEventAddError(startupEvent, error);
                error = nil;
            }

            [persistentContainer loadPersistentStoresWithCompletionHandler:^(NSPersistentStoreDescription *description, NSError *error) {
                SentryStartupEventAddError(startupEvent, error);
                [startupEvent sendWithTitle:@"Can't initialize Core Data stack"];

                @throw [NSException exceptionWithName:@"Can't initialize Core Data stack"
                                               reason:[error localizedDescription]
                                             userInfo:[error userInfo]];
            }];

        }
    }];
    return persistentContainer;
}

- (NSPersistentStoreDescription *)storeDescription
{
    if(_storeDescription) {
        return _storeDescription;
    }

    NSPersistentStoreDescription *storeDescription = [[NSPersistentStoreDescription alloc] initWithURL:self.storeURL];
    storeDescription.shouldInferMappingModelAutomatically = true;
    storeDescription.shouldMigrateStoreAutomatically = true;
    
    if ([self isSharedLoginEnabled]) {
        [storeDescription setOption:@YES forKey:NSPersistentHistoryTrackingKey];
        [storeDescription setOption:@YES forKey:NSPersistentStoreRemoteChangeNotificationPostOptionKey];
    }
    
    return storeDescription;
}

- (NSManagedObjectModel *)managedObjectModel
{
    if (_managedObjectModel) {
        return _managedObjectModel;
    }
    NSString *modelPath = [self modelPath];
    NSURL *modelURL = [NSURL fileURLWithPath:modelPath];
    if ([self.modelName isEqualToString:ContextManagerModelNameCurrent]) {
        // Use the current version defined in data model.
        _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
    } else {
        // Find the specific version defined in data model.
        NSURL *versionedModelURL = [[modelURL URLByAppendingPathComponent:self.modelName] URLByAppendingPathExtension:@"mom"];
        NSAssert([NSFileManager.defaultManager fileExistsAtPath:versionedModelURL.path],
                 @"Can't find model '%@' at %@", self.modelName, modelPath);
        _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:versionedModelURL];
    }
    return _managedObjectModel;
}

- (NSOperationQueue *)remoteChangeQueue
{
    if (_remoteChangeQueue) {
        return _remoteChangeQueue;
    }
    
    _remoteChangeQueue = [[NSOperationQueue alloc] init];
    _remoteChangeQueue.maxConcurrentOperationCount = 1;
    return _remoteChangeQueue;
}

- (void)setHistoryToken:(NSPersistentHistoryToken *)historyToken
{
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:historyToken
                                         requiringSecureCoding:YES
                                                         error:nil];
    
    [[NSUserDefaults standardUserDefaults] setObject:data forKey:@"transactionHistoryToken"];
}

- (NSPersistentHistoryToken *)historyToken
{
    NSData *data = [[NSUserDefaults standardUserDefaults] objectForKey:@"transactionHistoryToken"];
    if (data) {
        NSPersistentHistoryToken *token = [NSKeyedUnarchiver unarchivedObjectOfClass:[NSPersistentHistoryToken class]
                                                                            fromData:data
                                                                               error:nil];
        return token;
    }
    
    return nil;
}


#pragma mark - Private Helpers

- (void)migrateDataModelsIfNecessary:(SentryStartupEvent *)sentryEvent
{
    if (![[NSFileManager defaultManager] fileExistsAtPath:[[self storeURL] path]]) {
        DDLogInfo(@"No store exists at URL %@.  Skipping migration.", [self storeURL]);
        return;
    }
    
    NSDictionary *metadata = [NSPersistentStoreCoordinator metadataForPersistentStoreOfType:NSSQLiteStoreType
                                                               URL:[self storeURL]
                                                           options:nil
                                                             error:nil];
    BOOL migrationNeeded = ![self.managedObjectModel isConfiguration:nil compatibleWithStoreMetadata:metadata];
    
    if (migrationNeeded) {
        DDLogWarn(@"Migration required for persistent store.");
        NSError *error = nil;
        NSArray *sortedModelNames = [self sortedModelNames];

        [CoreDataIterativeMigrator iterativeMigrateWithSourceStore:[self storeURL]
        storeType:NSSQLiteStoreType
               to:self.managedObjectModel
            using:sortedModelNames
            error:&error];

        if (error != nil) {
            DDLogError(@"Unable to migrate store: %@", error);

            SentryStartupEventAddError(sentryEvent, error);
        }
    }
}

- (NSArray *)sortedModelNames
{
    NSString *modelPath = [self modelPath];
    NSString *versionPath = [modelPath stringByAppendingPathComponent:@"VersionInfo.plist"];
    NSDictionary *versionInfo = [NSDictionary dictionaryWithContentsOfFile:versionPath];
    NSArray *modelNames = [[versionInfo[@"NSManagedObjectModel_VersionHashes"] allKeys] sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
        return [obj1 compare:obj2 options:NSNumericSearch];
    }];
    
    return modelNames;
}

- (NSString *)modelPath
{
    return [[NSBundle mainBundle] pathForResource:@"WordPress" ofType:@"momd"];
}

- (BOOL)isSharedLoginEnabled
{
   return [Feature enabled:FeatureFlagSharedLogin];
}

- (NSURL *)localDatabasePath
{
    NSString *documentsDirectory = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory,
                                                                        NSUserDomainMask,
                                                                        YES) lastObject];
    return [NSURL fileURLWithPath:[documentsDirectory stringByAppendingPathComponent:@"WordPress.sqlite"]];
}

- (NSURL *)sharedDatabasePath
{
    NSURL *sharedDirectory = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:@"group.org.wordpress"];
    return [sharedDirectory URLByAppendingPathComponent:@"WordPress.sqlite"];
}

- (void)copyDatabaseToSharedDirectoryIfNeeded
{
    BOOL sharedDatabaseExists = [[NSFileManager defaultManager] fileExistsAtPath:[[self sharedDatabasePath] absoluteString]];
    if (![self isSharedLoginEnabled] || sharedDatabaseExists || AppConfiguration.isJetpack) {
        return;
    }
    
    NSString *shmLocalFilePath = [[[self localDatabasePath] absoluteString] stringByAppendingString:@"-shm"];
    NSString *walLocalFilePath = [[[self localDatabasePath] absoluteString] stringByAppendingString:@"-wal"];
    NSString *shmSharedFilePath = [[[self sharedDatabasePath] absoluteString] stringByAppendingString:@"-shm"];
    NSString *walSharedFilePath = [[[self sharedDatabasePath] absoluteString] stringByAppendingString:@"-wal"];
    [[NSFileManager defaultManager] copyItemAtURL:[self localDatabasePath] toURL:[self sharedDatabasePath] error:nil];
    [[NSFileManager defaultManager] copyItemAtURL:[NSURL URLWithString:shmLocalFilePath] toURL:[NSURL URLWithString:shmSharedFilePath] error:nil];
    [[NSFileManager defaultManager] copyItemAtURL:[NSURL URLWithString:walLocalFilePath] toURL:[NSURL URLWithString:walSharedFilePath] error:nil];
}

@end

File: WordPress/Classes/Utility/ContextManager.swift
import Foundation
import CoreData

/// A constant representing the current version of the data model.
///
/// - SeeAlso: ContextManager.init(modelName:store:)
let ContextManagerModelNameCurrent = "$CURRENT"

public protocol CoreDataStackSwift: CoreDataStack {

    /// Execute the given block with a background context and save the changes.
    ///
    /// This function _does not block_ its running thread. The block is executed in background and its return value
    /// is passed onto the `completion` block which is executed on the given `queue`.
    ///
    /// - Parameters:
    ///   - block: A closure which uses the given `NSManagedObjectContext` to make Core Data model changes.
    ///   - completion: A closure which is called with the return value of the `block`, after the changed made
    ///         by the `block` is saved.
    ///   - queue: A queue on which to execute the completion block.
    func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) -> T, completion: ((T) -> Void)?, on queue: DispatchQueue)

    /// Execute the given block with a background context and save the changes _if the block does not throw an error_.
    ///
    /// This function _does not block_ its running thread. The block is executed in background and the return value
    /// (or an error) is passed onto the `completion` block which is executed on the given `queue`.
    ///
    /// - Parameters:
    ///   - block: A closure that uses the given `NSManagedObjectContext` to make Core Data model changes. The changes
    ///         are only saved if the block does not throw an error.
    ///   - completion: A closure which is called with the `block`'s execution result, which is either an error thrown
    ///         by the `block` or the return value of the `block`.
    ///   - queue: A queue on which to execute the completion block.
    func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) throws -> T, completion: ((Result<T, Error>) -> Void)?, on queue: DispatchQueue)

    /// Execute the given block with a background context and save the changes _if the block does not throw an error_.
    ///
    /// - Parameter block: A closure that uses the given `NSManagedObjectContext` to make Core Data model changes.
    ///     The changes are only saved if the block does not throw an error.
    /// - Returns: The value returned by the `block`
    /// - Throws: The error thrown by the `block`, in which case the Core Data changes made by the `block` is discarded.
    func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) throws -> T) async throws -> T

}

@objc
public class ContextManager: NSObject, CoreDataStack, CoreDataStackSwift {
    static var inMemoryStoreURL: URL {
        URL(fileURLWithPath: "/dev/null")
    }

    private let modelName: String
    private let storeURL: URL
    private let persistentContainer: NSPersistentContainer

    /// A serial queue used to ensure there is only one writing operation at a time.
    ///
    /// - Note: This queue currently is not used in `performAndSave(_:)` the "save synchronously" function, since it's
    ///   not safe to block current thread. Considering the aforementioned `performAndSave(_:)` function is going to be
    ///   removed soon, I think it's okay to make this compromise.
    private let writerQueue: OperationQueue

    @objc
    public var mainContext: NSManagedObjectContext {
        persistentContainer.viewContext
    }

    convenience override init() {
        self.init(modelName: ContextManagerModelNameCurrent, store: Self.localDatabasePath)
    }

    /// Create a ContextManager instance with given model name and database location.
    ///
    /// Note: This initialiser is only used for testing purpose at the moment.
    ///
    /// - Parameters:
    ///   - modelName: Model name in Core Data data model file.
    ///         Use ContextManagerModelNameCurrent for current version, or
    ///         "WordPress <version>" for specific version.
    ///   - store: Database location. Use `ContextManager.inMemoryStoreURL` to create an in-memory database.
    init(modelName: String, store storeURL: URL) {
        assert(modelName == ContextManagerModelNameCurrent || modelName.hasPrefix("WordPress "))
        assert(storeURL.isFileURL)

        self.modelName = modelName
        self.storeURL = storeURL
        self.persistentContainer = Self.createPersistentContainer(storeURL: storeURL, modelName: modelName)
        self.writerQueue = OperationQueue()
        self.writerQueue.name = "org.wordpress.CoreDataStack.writer"
        self.writerQueue.maxConcurrentOperationCount = 1

        super.init()

        mainContext.automaticallyMergesChangesFromParent = true
        mainContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
        NullBlogPropertySanitizer(context: mainContext).sanitize()
    }

    public func newDerivedContext() -> NSManagedObjectContext {
        let context = persistentContainer.newBackgroundContext()
        context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
        return context
    }

    @objc(performAndSaveUsingBlock:)
    public func performAndSave(_ block: @escaping (NSManagedObjectContext) -> Void) {
        let context = newDerivedContext()
        context.performAndWait {
            block(context)

            self.save(context, .alreadyInContextQueue)
        }
    }

    @objc(performAndSaveUsingBlock:completion:onQueue:)
    public func performAndSave(_ block: @escaping (NSManagedObjectContext) -> Void, completion: (() -> Void)?, on queue: DispatchQueue) {
        let context = newDerivedContext()
        self.writerQueue.addOperation(AsyncBlockOperation { done in
            context.perform {
                block(context)

                self.save(context, .alreadyInContextQueue)
                queue.async { completion?() }
                done()
            }
        })
    }

    public func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) throws -> T, completion: ((Result<T, Error>) -> Void)?, on queue: DispatchQueue) {
        let context = newDerivedContext()
        self.writerQueue.addOperation(AsyncBlockOperation { done in
            context.perform {
                let result = Result(catching: { try block(context) })
                if case .success = result {
                    self.save(context, .alreadyInContextQueue)
                }
                queue.async { completion?(result) }
                done()
            }
        })
    }

    public func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) -> T, completion: ((T) -> Void)?, on queue: DispatchQueue) {
        performAndSave(
            block,
            completion: { (result: Result<T, Error>) in
                // It's safe to force-unwrap here, since the `block` does not throw an error.
                completion?(try! result.get())
            },
            on: queue
        )
    }

    public func performAndSave<T>(_ block: @escaping (NSManagedObjectContext) throws -> T) async throws -> T {
        try await withCheckedThrowingContinuation { continuation in
            performAndSave(block, completion: continuation.resume(with:), on: DispatchQueue.global())
        }
    }

    @objc
    public func saveContextAndWait(_ context: NSManagedObjectContext) {
        save(context, .synchronously)
    }

    @objc(saveContext:)
    public func save(_ context: NSManagedObjectContext) {
        save(context, .asynchronously)
    }

    @objc(saveContext:withCompletionBlock:onQueue:)
    public func save(_ context: NSManagedObjectContext, completion: (() -> Void)?, on queue: DispatchQueue) {
        if let completion {
            save(context, .asynchronouslyWithCallback(completion: completion, queue: queue))
        } else {
            save(context, .asynchronously)
        }
    }

    static func migrateDataModelsIfNecessary(storeURL: URL, objectModel: NSManagedObjectModel) throws {
        guard FileManager.default.fileExists(atPath: storeURL.path) else {
            DDLogInfo("No store exists at \(storeURL).  Skipping migration.")
            return
        }

        guard let metadata = try? NSPersistentStoreCoordinator.metadataForPersistentStore(ofType: NSSQLiteStoreType, at: storeURL),
            !objectModel.isConfiguration(withName: nil, compatibleWithStoreMetadata: metadata)
        else {
            return
        }

        DDLogWarn("Migration required for persistent store.")

        guard let modelFileURL = Bundle.main.url(forResource: "WordPress", withExtension: "momd") else {
            fatalError("Can't find WordPress.momd")
        }

        guard let versionInfo = NSDictionary(contentsOf: modelFileURL.appendingPathComponent("VersionInfo.plist")) else {
            fatalError("Can't get the object model's version info")
        }

        guard let modelNames = (versionInfo["NSManagedObjectModel_VersionHashes"] as? [String: AnyObject])?.keys else {
            fatalError("Can't parse the model versions")
        }

        let sortedModelNames = modelNames.sorted { $0.compare($1, options: .numeric) == .orderedAscending }
        try CoreDataIterativeMigrator.iterativeMigrate(
            sourceStore: storeURL,
            storeType: NSSQLiteStoreType,
            to: objectModel,
            using: sortedModelNames
        )
    }
}

// MARK: - Private methods

private extension ContextManager {
    static var localDatabasePath: URL {
        guard let url = try? FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true) else {
            fatalError("Failed to find the document directory")
        }

        return url.appendingPathComponent("WordPress.sqlite")
    }

    func save(_ context: NSManagedObjectContext, _ option: SaveContextOption) {
        let block: () -> Void = {
            self.internalSave(context)

            switch option {
            case let .asynchronouslyWithCallback(completion, queue):
                queue.async(execute: completion)
            case .synchronously, .asynchronously, .alreadyInContextQueue:
                // Do nothing
                break
            }
        }

        // Ensure that the `context`'s concurrency type is not `confinementConcurrencyType`, since it will crash if `perform` or `performAndWait` is called.
        guard context.concurrencyType == .mainQueueConcurrencyType || context.concurrencyType == .privateQueueConcurrencyType else {
            block()
            return
        }

        switch option {
        case .synchronously:
            context.performAndWait(block)
        case .asynchronously, .asynchronouslyWithCallback:
            context.perform(block)
        case .alreadyInContextQueue:
            block()
        }
    }
}

// MARK: - Initialise Core Data stack

private extension ContextManager {
    static func createPersistentContainer(storeURL: URL, modelName: String) -> NSPersistentContainer {
        guard var modelFileURL = Bundle.main.url(forResource: "WordPress", withExtension: "momd") else {
            fatalError("Can't find WordPress.momd")
        }

        if modelName != ContextManagerModelNameCurrent {
            modelFileURL = modelFileURL.appendingPathComponent(modelName).appendingPathExtension("mom")
        }

        guard let objectModel = NSManagedObjectModel(contentsOf: modelFileURL) else {
            fatalError("Can't create object model named \(modelName) at \(modelFileURL)")
        }

        let startupEvent = SentryStartupEvent()

        do {
            try migrateDataModelsIfNecessary(storeURL: storeURL, objectModel: objectModel)
        } catch {
            DDLogError("Unable to migrate store: \(error)")
            startupEvent.add(error: error as NSError)
        }

        let storeDescription = NSPersistentStoreDescription(url: storeURL)
        storeDescription.shouldInferMappingModelAutomatically = true
        storeDescription.shouldMigrateStoreAutomatically = true
        let persistentContainer = NSPersistentContainer(name: "WordPress", managedObjectModel: objectModel)
        persistentContainer.persistentStoreDescriptions = [storeDescription]
        persistentContainer.loadPersistentStores { _, error in
            guard let error else {
                return
            }

            DDLogError("Error opening the database. \(error)\nDeleting the file and trying again")
            startupEvent.add(error: error)

            // make a backup of the old database
            do {
                try CoreDataIterativeMigrator.backupDatabase(at: storeURL)
            } catch {
                startupEvent.add(error: error)
            }

            startupEvent.send(title: "Can't initialize Core Data stack")
            objc_exception_throw(
                NSException(
                    name: NSExceptionName(rawValue: "Can't initialize Core Data stack"),
                    reason: error.localizedDescription,
                    userInfo: (error as NSError).userInfo
                )
            )
        }

        return persistentContainer
    }

}

extension ContextManager {
    private static let internalSharedInstance = ContextManager()
    /// Tests purpose only
    static var overrideInstance: ContextManager?

    @objc class func sharedInstance() -> ContextManager {
        if let overrideInstance = overrideInstance {
            return overrideInstance
        }

        return ContextManager.internalSharedInstance
    }

    static var shared: ContextManager {
        return sharedInstance()
    }
}

private enum SaveContextOption {
    case synchronously
    case asynchronously
    case asynchronouslyWithCallback(completion: () -> Void, queue: DispatchQueue)
    case alreadyInContextQueue
}

/// Use this temporary workaround to mitigate Core Data concurrency issues when accessing the given `object`.
///
/// When the app is launched from Xcode, some code may crash due to the effect of the "com.apple.CoreData.ConcurrencyDebug"
/// launch argument. The crash indicates the crash site violates [the following rule](https://developer.apple.com/documentation/coredata/using_core_data_in_the_background#overview)
///
/// > To use Core Data in a multithreaded environment, ensure that:
/// > - Managed objects retrieved from a context are bound to the same queue that the context is bound to.
///
/// This function can be used as a temporary workaround to mitigate aforementioned crashes during development.
///
/// - Warning: The workaround does not apply to release builds. In a release build, calling this function is exactly
///     the same as calling the given `closure` directly.
///
/// - Warning: This function is _not_ a solution for Core Data concurrency issues, and should only be used as a
///     temporary solution, to avoid the Core Data concurrency issue becoming a blocker to feature developlement.
@available(*, deprecated, message: "This workaround is meant as a temporary solution to mitigate Core Data concurrency issues when accessing the `object`. Please see this function's API doc for details.")
@inlinable
public func workaroundCoreDataConcurrencyIssue<Value>(accessing object: NSManagedObject, _ closure: () -> Value) -> Value {
#if DEBUG
    guard let context = object.managedObjectContext else {
        fatalError("The object must be bound to a context: \(object)")
    }

    var value: Value!
    context.performAndWait {
        value = closure()
    }
    return value
#else
    return closure()
#endif /* DEBUG */
}
